### PR TITLE
RD-2153 Added display_name support to blueprint deploy modal

### PR DESCRIPTION
**Pull request by kubama**

### CODE REVIEW

**Comment by qooban** [comment]
I'll review it tomorrow morning. 

BTW There are some issues in the build:

<details>
  <summary>Click to see errors</summary>

```
[2021-05-12T10:59:46.349Z]     ERROR in widgets/common/src/DeployBlueprintModal.tsx:26:41
[2021-05-12T10:59:46.349Z]     prettier/prettier: Replace `deploymentId:·string,·{·deploymentName:·deploymentName·}:·BlueprintDeployParams` with `⏎········deploymentId:·string,⏎········{·deploymentName:·deploymentName·}:·BlueprintDeployParams⏎····`
[2021-05-12T10:59:46.349Z]         24 |     }
[2021-05-12T10:59:46.349Z]         25 | 
[2021-05-12T10:59:46.349Z]       > 26 |     function waitForDeploymentIsCreated(deploymentId: string, { deploymentName: deploymentName }: BlueprintDeployParams) {
[2021-05-12T10:59:46.349Z]            |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2021-05-12T10:59:46.349Z]         27 |         const deploymentActions = new DeploymentActions(toolbox);
[2021-05-12T10:59:46.349Z]         28 | 
[2021-05-12T10:59:46.349Z]         29 |         return deploymentActions
[2021-05-12T10:59:46.349Z]     
[2021-05-12T10:59:46.349Z]     ERROR in widgets/common/src/DeployBlueprintModal.tsx:26:65
[2021-05-12T10:59:46.349Z]     no-useless-rename: Destructuring assignment deploymentName unnecessarily renamed.
[2021-05-12T10:59:46.349Z]         24 |     }
[2021-05-12T10:59:46.349Z]         25 | 
[2021-05-12T10:59:46.349Z]       > 26 |     function waitForDeploymentIsCreated(deploymentId: string, { deploymentName: deploymentName }: BlueprintDeployParams) {
[2021-05-12T10:59:46.349Z]            |                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2021-05-12T10:59:46.349Z]         27 |         const deploymentActions = new DeploymentActions(toolbox);
[2021-05-12T10:59:46.349Z]         28 | 
[2021-05-12T10:59:46.349Z]         29 |         return deploymentActions
[2021-05-12T10:59:46.349Z]     
[2021-05-12T10:59:46.349Z]     ERROR in widgets/common/src/DeployBlueprintModal.tsx:37:11
[2021-05-12T10:59:46.349Z]     no-useless-rename: Destructuring assignment deploymentName unnecessarily renamed.
[2021-05-12T10:59:46.349Z]         35 |     function installDeployment(
[2021-05-12T10:59:46.350Z]         36 |         deploymentId: string,
[2021-05-12T10:59:46.350Z]       > 37 |         { deploymentName: deploymentName }: BlueprintDeployParams,
[2021-05-12T10:59:46.350Z]            |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2021-05-12T10:59:46.350Z]         38 |         installWorkflowParameters: Record<string, any>,
[2021-05-12T10:59:46.350Z]         39 |         installWorkflowOptions: WorkflowOptions
[2021-05-12T10:59:46.350Z]         40 |     ) {
```
</details>

File: test/cypress/integration/widgets/deployment_button_spec.js [hunk 1]
```diff
@@ -189,7 +192,7 @@ describe('Create Deployment Button widget', () => {
             cy.get('div.deployBlueprintModal div.error.message').within(() => {
                 cy.get('li:nth-child(1)').should(
                     'have.text',
-                    `Deployment ${deploymentName} installation failed: Cannot start install workflow`
+                    `Deployment ${deploymentId} installation failed: Cannot start install workflow`
```
Comment (Gelio) [inline test/cypress/integration/widgets/deployment_button_spec.js:195]: Do we not want to validate that it's the `deploymentName` that's used in the error message? We could easily do I hope by having a different deployment name and id, as in other tests

File: test/cypress/integration/widgets/deployment_button_spec.js [hunk 2]
```diff
@@ -201,13 +204,10 @@ describe('Create Deployment Button widget', () => {
...
-            cy.get(`form :nth-child(${firstInputNthChild}).field`)
-                .as('string_no_default')
-                .within(() => {
-                    cy.get('input').clear().type('Something');
-                });
+            cy.get('input[name="deploymentId"]').clear().type(deploymentName);
+            cy.get('input[name=string_no_default]').clear().type('Something');
 
-            cy.get(`form :nth-child(${firstInputNthChild + 1}).field`)
+            cy.contains('.field', 'string_constraint_pattern')
```
Comment (Gelio) [inline test/cypress/integration/widgets/deployment_button_spec.js:210]: Can you tell me what this is :smile: ? I don't know what's happening here. `string_no_default` field name :thinking: ?

**Review by Gelio — DISMISSED** [review]
Having a video/screenshots of the functionality would make the PR easier to review :slightly_smiling_face: 

Generally the implementation looks sound :+1: I did not test it in the UI myself

File: test/cypress/integration/widgets/deployment_button_spec.js (hunk 2 above)
Comment (kubama) [reply test/cypress/integration/widgets/deployment_button_spec.js:210]: It's a blueprint input field. I think referring it by its name (thus by the filed name) is a more robust way than using filed index, so I changed it here as it broke after adding deployment name input.

File: test/cypress/integration/widgets/deployment_button_spec.js (hunk 1 above)
Comment (kubama) [reply test/cypress/integration/widgets/deployment_button_spec.js]: Right, updated

File: test/cypress/integration/widgets/deployment_button_spec.js (hunk 2 above)
Comment (Gelio) [reply test/cypress/integration/widgets/deployment_button_spec.js:211]: Makes sense now :smile: I thought we had such a hardcoded field in the UI

**Comment by Gelio** [comment]
Interesting, the PR is still approved despite pushing a new commit to it :smile: Anyway, looks good to me, so I will not reapprove

File: test/cypress/integration/widgets/deployment_button_spec.js [hunk 3]
```diff
@@ -201,13 +205,10 @@ describe('Create Deployment Button widget', () => {
...
```
Comment (qooban) [inline test/cypress/integration/widgets/deployment_button_spec.js:211]: Optional:

As you did it in one place, did you consider also changing it in other places:
```
cy.get(`form :nth-child(${firstInputNthChild + x}).field`)
    .as(fieldName)
```
to
```
cy.contains('.field', fieldName)
```
?

There's around 20 occurrences if I'm not mistaken.

File: widgets/common/src/GenericDeployModal.jsx [hunk 4]
```diff
@@ -212,16 +216,21 @@ class GenericDeployModal extends React.Component {
...
 
-            if (showDeploymentNameInput && _.isEmpty(deploymentName)) {
-                errors.deploymentName = t('errors.noDeploymentName');
+            if (showDeploymentNameInput) {
+                if (_.isEmpty(deploymentId)) {
+                    errors.deploymentId = t('errors.noDeploymentId');
+                }
+                if (_.isEmpty(deploymentName)) {
+                    errors.deploymentName = t('errors.noDeploymentName');
+                }
```
Comment (qooban) [inline widgets/common/src/GenericDeployModal.jsx:233]: Optional:

Probably more accurate would be to report errors in the same order as the fields are presented.
Right now we have:
image

If you change the conditions to the following order: 1. `deploymentName`, 2. `deploymentId`, 3. `blueprintName`, then IMHO it'll look a bit better.

File: test/cypress/support/deployments.ts [hunk 5]
```diff
@@ -66,6 +66,7 @@ const commands = {
         cy.get('.deploymentsWidget').within(() => {
             cy.getSearchInput().clear().type(deploymentId);
             cy.get('.input.loading').should('not.exist');
+            cy.get('.widgetLoader').should('be.not.visible');
```
Comment (qooban) [inline test/cypress/support/deployments.ts:69]: Out of curiosity: Is it due to the code change coming from the PR or just an addition to make the test less error-prone?

**Review by qooban — DISMISSED** [review]
Looks good 👍🏻  Added some optional comments.

Did UI tests, so it seems all working fine. 

Looking at the final story for 6.0 - RD-2151 related to display name, I'm wondering:
- if breadcrumb changes shouldn't be done as a separate story as it is not strictly connected with Deployment View widget (I'm afraid it might be a bit tricky, I'm not sure, just for you to consider)
- if any other alignments are planned? (I've listed what I found in [this comment](https://cloudifysource.atlassian.net/wiki/spaces/WR/pages/1769504860/Deployment+display+name?focusedCommentId=1784709188#comment-1784709188))

File: test/cypress/support/deployments.ts (hunk 5 above)
Comment (kubama) [reply test/cypress/support/deployments.ts:69]: I wouldn't say it's related to the changes from this PR, but still, `filter_spec` started to fail on this development branch so I added this additional check as a fix.

File: widgets/common/src/GenericDeployModal.jsx (hunk 4 above)
Comment (kubama) [reply widgets/common/src/GenericDeployModal.jsx]: Good suggestion - fixed

File: test/cypress/integration/widgets/deployment_button_spec.js (hunk 3 above)
Comment (kubama) [reply test/cypress/integration/widgets/deployment_button_spec.js:210]: Updated

**Comment by kubama** [comment]
> * if breadcrumb changes shouldn't be done as a separate story as it is not strictly connected with Deployment View widget (I'm afraid it might be a bit tricky, I'm not sure, just for you to consider)

I'm not sure, I'll decide as I go.

> * if any other alignments are planned? (I've listed what I found in [this comment](https://cloudifysource.atlassian.net/wiki/spaces/WR/pages/1769504860/Deployment+display+name?focusedCommentId=1784709188#comment-1784709188))

Looks like there are areas identified in the comment you are referring to that are not covered by RD-2151. Not sure if/how they should be handled.

**Review by qooban — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**